src/main/java/com/commercetools/sync/sdk2/products/helpers/PriceReferenceResolver.java

```diff
@@ -0,0 +1,296 @@
+                              .whenComplete(
+                                  (draftWithCreatedChannel, exception) -> {
+                                    if (exception != null) {
+                                      result.completeExceptionally(
+                                          new ReferenceResolutionException(
+                                              format(
+                                                  FAILED_TO_RESOLVE_REFERENCE,
+                                                  ChannelReference.CHANNEL,
+                                                  draftBuilder.getCountry(),
+                                                  draftBuilder.getValue(),
+                                                  exception.getMessage()),
+                                              exception));
+                                    } else {
+                                      result.complete(draftWithCreatedChannel);
+                                    }
+                                  });
+                          return result;
+                        }));
+  }
+
+  /**
+   * Helper method that creates a new {@link io.sphere.sdk.channels.Channel} on the CTP project with
+   * the specified {@code channelKey} and of the role {@code "InventorySupply"}. Only if the {@code
+   * ensureChannels} options is set to {@code true} on the {@code options} instance of {@code this}
+   * class. Then it resolves the supply channel reference on the supplied {@code
+   * inventoryEntryDraft} by setting the id of it's supply channel reference with the newly created
+   * Channel.
+   *
+   * <p>If the {@code ensureChannels} options is set to {@code false} on the {@code options}
+   * instance of {@code this} class, the future is completed exceptionally with a {@link
+   * ReferenceResolutionException}.
+   *
+   * <p>The method then returns a CompletionStage with a resolved channel reference {@link
+   * io.sphere.sdk.products.PriceDraftBuilder} object.
+   *
+   * @param channelKey the key to create the new channel with.
+   * @param draftBuilder the inventory entry draft builder where to resolve it's supply channel
+   *     reference to the newly created channel.
+   * @return a CompletionStage with the same {@code draftBuilder} instance having resolved channel
+   *     reference.
+   */
+  @Nonnull
+  private CompletionStage<PriceDraftBuilder> createChannelAndSetReference(
+      @Nonnull final String channelKey, @Nonnull final PriceDraftBuilder draftBuilder) {
+
+    if (options.shouldEnsurePriceChannels()) {
+      return channelService
+          .createAndCacheChannel(channelKey)
+          .thenCompose(
+              createdChannelOptional -> {
+                if (createdChannelOptional.isPresent()) {
+                  return completedFuture(
+                      draftBuilder.channel(createdChannelOptional.get().toResourceIdentifier()));
```

Comment:
Note that `toResourceIdentifier()` doesn't set the key. Might be better to use `ChannelResourceIdentifierBuilder.of().key(createdChannelOptional.get().getKey()).build();` instead.

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/931/commits/53fb759945b160f8e8ad9083e7f56216bda984e8